ring/ownership_priority_queue.go

```diff
@@ -0,0 +1,200 @@
+package ring
+
+import (
+	"container/heap"
+	"fmt"
+	"math"
+)
+
+type ringItem interface {
+	GetID() int
+	Less(other ringItem) bool
+}
+
+type ringInstance struct {
+	instanceID int
+}
+
+func newRingInstance(instanceID int) *ringInstance {
+	return &ringInstance{
+		instanceID: instanceID,
+	}
+}
+
+func (ri *ringInstance) GetID() int {
+	return ri.instanceID
+}
+
+func (ri *ringInstance) Less(other ringItem) bool {
+	ringInstance, ok := other.(*ringInstance)
+	if !ok {
+		return false
+	}
+	return ri.instanceID < ringInstance.instanceID
+}
+
+func (ri *ringInstance) String() string {
+	return fmt.Sprintf("[instanceID: %d]", ri.instanceID)
+}
+
+type ringToken struct {
+	token     uint32
+	prevToken uint32
+}
+
+func newRingToken(token, prevToken uint32) *ringToken {
+	return &ringToken{
+		token:     token,
+		prevToken: prevToken,
+	}
+}
+
+func (rt *ringToken) GetID() int {
+	return int(rt.token)
+}
+
+func (rt *ringToken) Less(other ringItem) bool {
+	ringToken, ok := other.(*ringToken)
+	if !ok {
+		return false
+	}
+	return rt.token < ringToken.token
+}
+
+func (rt *ringToken) String() string {
+	return fmt.Sprintf("[token: %d, prevToken: %d]", rt.token, rt.prevToken)
+}
+
+type ownershipInfo struct {
+	ringItem  ringItem
+	ownership float64
+	index     int
+}
+
+func newRingTokenOwnershipInfo(token, prevToken uint32) *ownershipInfo {
+	rT := newRingToken(token, prevToken)
+	ownership := float64(getTokenDistance(prevToken, token))
+	return &ownershipInfo{
+		ownership: ownership,
+		ringItem:  rT,
+	}
+}
+
+func newRingInstanceOwnershipInfo(instanceID int, ownership float64) *ownershipInfo {
+	rI := newRingInstance(instanceID)
+	return &ownershipInfo{
+		ownership: ownership,
+		ringItem:  rI,
+	}
+}
+
+// ownershipPriorityQueue is a max-heap, i.e., a priority queue
+// where items with a higher priority will be extracted first.
+// Namely, items with a higher ownership have a higher priority.
+// In order to guarantee that 2 instances of ownershipPriorityQueue
+// with the same items always assign equal priorities to equal items,
+// in the case of items with equal ownership, we rely on the
+// order of item ids.
+type ownershipPriorityQueue struct {
+	items []*ownershipInfo
```

Comment:
Given that one `ownershipPriorityQueue` should contain homogeneous elements (it can't contain some instances and some tokens), I think this should be:
```go
type ringItem[T any] interface {
	GetID() int
	Less(other T) bool
}

type ownershipInfo[T ringItem[T]] struct {
	ringItem  T
	ownership float64
	index     int
}

type ownershipPriorityQueue[T ringItem[T]] struct {
	items []*ownershipInfo[T]
}

func newPriorityQueue[T ringItem[T]](len int) *ownershipPriorityQueue[T] {
	return &ownershipPriorityQueue[T]{
		items: make([]*ownershipInfo[T], 0, len),
	}
}
```

This way the specific implementations don't need to type assert, and they can be just:

```go
type ringInstance struct{ instanceID int }

func (ri *ringInstance) GetID() int                    { return ri.instanceID }
func (ri *ringInstance) Less(other *ringInstance) bool { return ri.instanceID < other.instanceID }

type ringToken struct{ token, prevToken uint32 }

func (rt *ringToken) GetID() int                 { return int(rt.token) }
func (rt *ringToken) Less(other *ringToken) bool { return rt.token < other.token }
```

Reply:
While reviewing the code further, I realized that then we also don't need the `GetID()` method. And we also don't need any of those methods exporter. All we need need is to compare the values. However that can be made even easier.

```go
// ordered is like comparable constraint but it returns an int key to compare instead of being comparable itself.
type ordered interface{ key() int }

type ringInstance struct{ instanceID int }

func (ri ringInstance) key() int       { return ri.instanceID }

type ringToken struct{ token, prevToken uint32 }

func (rt ringToken) key() int { return int(rt.token) }

// ...


type ownershipInfo[T ordered] struct {
	ringItem  T
	ownership float64
	index     int
}


// ...

type ownershipPriorityQueue[T ordered] struct {
	items []*ownershipInfo[T]
}

func newPriorityQueue[T ordered](len int) *ownershipPriorityQueue[T] {
	return &ownershipPriorityQueue[T]{
		items: make([]*ownershipInfo[T], 0, len),
	}
}

```